src/main/scala/com/wavesplatform/state2/NgState.scala

```diff
@@ -1,38 +1,69 @@
 package com.wavesplatform.state2
 
+import java.util.concurrent.TimeUnit
+
+import cats.kernel.Monoid
+import com.google.common.cache.CacheBuilder
 import scorex.block.Block.BlockId
 import scorex.block.{Block, MicroBlock}
 import scorex.transaction.History.BlockMinerInfo
-import scorex.transaction.{AssetId, DiscardedMicroBlocks, Transaction}
+import scorex.transaction.{DiscardedMicroBlocks, Transaction}
+
+import scala.collection.mutable.{ListBuffer => MList, Map => MMap}
+
+class NgState(val base: Block, val baseBlockDiff: BlockDiff, val acceptedFeatures: Set[Short]) {
+
+  private val MaxTotalDiffs = 3
 
+  private val microDiffs: MMap[BlockId, (BlockDiff, Long)] = MMap.empty
+  private val micros: MList[MicroBlock] = MList.empty // fresh head
+  private val totalBlockDiffCache = CacheBuilder.newBuilder()
+    .maximumSize(MaxTotalDiffs)
+    .expireAfterWrite(10, TimeUnit.MINUTES)
+    .build[BlockId, BlockDiff]()
 
-case class NgState private(base: Block, diffs: Map[BlockId, (BlockDiff, Long)], micros: List[MicroBlock], acceptedFeatures: Set[Short]) {
 
-  lazy val lastMicroTotalSig: Option[ByteStr] =
-    micros.headOption.map(_.totalResBlockSig)
+  def microBlockIds: Seq[BlockId] = micros.map(_.totalResBlockSig).toList
+
+  private def diffFor(totalResBlockSig: BlockId): BlockDiff =
+    if (totalResBlockSig == base.uniqueId)
+      baseBlockDiff
+    else Option(totalBlockDiffCache.getIfPresent(totalResBlockSig)) match {
+      case Some(d) => d
+      case None =>
+        val prevResBlockSig = micros.find(_.totalResBlockSig == totalResBlockSig).get.prevResBlockSig
```

Comment:
Can we miss a microblock with that `totalResBlockSig`?

Reply:
This is private method. if we fail with `get`, it's a programming error